### PR TITLE
Add recoverability behavior

### PR DESCRIPTION
**Pull request by mikeminutillo**
Connects to #41

### CODE REVIEW

**Comment by mikeminutillo** [comment]
Updated to use proper fork connectors. Now it looks like this:

image

File: src/NServiceBus.Gateway.Tests/APIApprovals.Approve.approved.txt [hunk 1]
```diff
@@ -130,6 +130,13 @@ namespace NServiceBus.Gateway.Channels
         public string Type { get; set; }
     }
 }
+namespace NServiceBus.Gateway.Recoverability
+{
+    public class static IncomingMessageExtensions
+    {
+        public static int GetCurrentRetry(this NServiceBus.Transports.IncomingMessage message) { }
```
Comment (mikeminutillo) [inline src/NServiceBus.Gateway.Tests/APIApprovals.Approve.approved.txt]: Does it make sense for the gateway to expose this extension method? Alternatively we pass in the retry number and the exception details into the retry policy.

**Comment by mikeminutillo** [comment]
@Particular/nservicebus-maintainers As the owners of this repo, could get this PR reviewed and merged please.

**Comment by janovesk** [comment]
No longer WIP, ready to go.

File: src/NServiceBus.Gateway.AcceptanceTests/NServiceBus.Gateway.AcceptanceTests.csproj [hunk 2]
```diff
@@ -36,15 +36,15 @@
...
+      <HintPath>..\packages\NServiceBus.AcceptanceTesting.6.0.0-alpha0875\lib\net452\NServiceBus.AcceptanceTesting.dll</HintPath>
       <Private>True</Private>
     </Reference>
     <Reference Include="NServiceBus.Callbacks, Version=1.0.0.0, Culture=neutral, PublicKeyToken=9fc386479f8a226c, processorArchitecture=MSIL">
       <HintPath>..\packages\NServiceBus.Callbacks.1.0.0-alpha0058\lib\net452\NServiceBus.Callbacks.dll</HintPath>
       <Private>True</Private>
     </Reference>
     <Reference Include="NServiceBus.Core, Version=6.0.0.0, Culture=neutral, PublicKeyToken=9fc386479f8a226c, processorArchitecture=MSIL">
-      <HintPath>..\packages\NServiceBus.6.0.0-beta0001\lib\net452\NServiceBus.Core.dll</HintPath>
+      <HintPath>..\packages\NServiceBus.6.0.0-alpha0875\lib\net452\NServiceBus.Core.dll</HintPath>
```
Comment (timbussmann) [inline src/NServiceBus.Gateway.AcceptanceTests/NServiceBus.Gateway.AcceptanceTests.csproj]: shouldn't we use the beta package?

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    public class When_sending_fails_with_retries : NServiceBusAcceptanceTest
+    {
+        [Test]
+        public async Task Should_have_been_retried()
+        {
+            await Scenario.Define<Context>(c =>
+            {
+                c.Id = Guid.NewGuid();
+            })
+            .WithEndpoint<HeadquartersWithDefaultRetryPolicy>(b => b.When(async (bus, context) => await bus.SendToSites(new[]{"SiteA"}, new AnyMessage{Id = context.Id})))
```
Comment (timbussmann) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs]: doesn't need to be async, just return the Task from bus.SendToSites

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+            .WithEndpoint<ErrorSpy>()
+            .Done(c => c.MessageMovedToErrorQueue)
+            .Repeat(r => r.For(Transports.Default))
```
Comment (timbussmann) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs]: is there any reason in using Repeat in downstream acceptance tests?

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        {
+            public HeadquartersWithCustomRetryPolicy()
+            {
+                EndpointSetup<DefaultServer>((c, d) =>
+                {
+                    c.ScaleOut().InstanceDiscriminator("1");
+                    c.EnableFeature<Gateway>();
+                    c.EnableFeature<TimeoutManager>();
+                    c.SendFailedMessagesTo(ErrorSpyQueueName);
+                    c.Gateway().CustomRetryPolicy((msg, ex, currentRetry) =>
```
Comment (timbussmann) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs]: is this the only difference to `HeadquartersWithDefaultRetryPolicy`? In that case you can set that up on ` .WithEndpoint` within the testcase itself.

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+                                        {
+                                             Address = "http://localhost:25999/Headquarters/",
+                                            ChannelType = "http"
+                                        }
+                                };
+                    });
+            }
+        }
+
+        [Serializable]
```
Comment (timbussmann) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs]: shouldn't be needed?

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_without_retries.cs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            await Scenario.Define<Context>(c =>
+            {
+                c.Id = Guid.NewGuid();
+            })
+            .WithEndpoint<Headquarters>(b => b.When(async (bus, context) => await bus.SendToSites(new[]{"SiteA"}, new AnyMessage{Id = context.Id})))
+            .WithEndpoint<ErrorSpy>()
+            .Done(c => c.MessageMovedToErrorQueue)
+            .Repeat(r => r.For(Transports.Default))
+            .Should(c => Assert.IsTrue(c.MessageMovedToErrorQueue))
+            .Run();
```
Comment (timbussmann) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_without_retries.cs]: I think the formatting would look different when the particular format is applied?

File: src/NServiceBus.Gateway.AcceptanceTests/NServiceBus.Gateway.AcceptanceTests.csproj (hunk 2 above)
Comment (danielmarbach) [inline src/NServiceBus.Gateway.AcceptanceTests/NServiceBus.Gateway.AcceptanceTests.csproj]: No unfortunately we have to do it currently this way. We didn't want to burn a beta version for Core. Updating the current beta is also out of question because that would require some tricky nuget caching workarounds.

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        public async Task Should_use_custom_retry_policy()
+        {
+            await Scenario.Define<Context>(c =>
+            {
+                c.Id = Guid.NewGuid();
+            })
+            .WithEndpoint<HeadquartersWithCustomRetryPolicy>(b => b.When(async (bus, context) => await bus.SendToSites(new[] { "SiteA" }, new AnyMessage { Id = context.Id })))
+            .WithEndpoint<ErrorSpy>()
+            .Done(c => c.MessageMovedToErrorQueue)
+            .Repeat(r => r.For(Transports.Default))
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs]: Remove repeat here

File: src/NServiceBus.Gateway/GatewayHttpListenerInstaller.cs [hunk 9]
```diff
@@ -12,6 +12,11 @@ class GatewayHttpListenerInstaller : INeedToInstallSomething
         readonly IManageReceiveChannels channelManager;
         readonly bool enabled;
 
+        public GatewayHttpListenerInstaller()
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/GatewayHttpListenerInstaller.cs]: Forgive my ignorance, is this related to this PR and why?

File: src/NServiceBus.Gateway/Routing/Sites/ConfigurationBasedSiteRouter.cs [hunk 10]
```diff
@@ -4,6 +4,13 @@ namespace NServiceBus.Gateway.Routing.Sites
 
     class ConfigurationBasedSiteRouter : IRouteMessagesToSites
     {
+        readonly IDictionary<string, Site> sites;
```
Comment (danielmarbach) [inline src/NServiceBus.Gateway/Routing/Sites/ConfigurationBasedSiteRouter.cs]: Can you guys update the resharper code styles and apply a format on this code. The field should be lower down. Readonly is not necessary,

File: src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+            string value;
+            if (message.Headers.TryGetValue(Headers.Retries, out value))
+            {
+                int i;
+                if (int.TryParse(value, out i))
+                {
+                    return i + 1;
+                }
+            }
+            return 1;
```
Comment (timbussmann) [inline src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs]: this seems weird and like an attempt to reimplement the configuration behavior from NSB 5 (e.g. MaxRetries=2 == 1 retry attempt). If there was no retry, returning 0 would make more sense to me?

File: src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+        private async Task<bool> SendForRetry(ITransportReceiveContext context, Exception exception, Func<IRoutingContext, Task> fork)
+        {
+            var message = context.Message;
+
+            var currentRetry = GetCurrentRetry(message);
+
+            var timeIncrease = retryPolicy(message, exception, currentRetry);
+
+            if (timeIncrease <= TimeSpan.Zero)
```
Comment (timbussmann) [inline src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs]: in the core, `TimeSpan.Zero` means immediate retry in the custom policy, where `TimeSpan.MinValue` means "do not retry". I think it would make sense to align these?

File: src/NServiceBus.Gateway/Routing/Sites/ConfigurationBasedSiteRouter.cs (hunk 10 above)
Comment (timbussmann) [inline src/NServiceBus.Gateway/Routing/Sites/ConfigurationBasedSiteRouter.cs]: apply particular code layout rules

File: src/NServiceBus.Gateway/Settings/GatewaySettings.cs [hunk 13]
```diff
@@ -36,5 +37,38 @@ public void ChannelFactories(Func<string, IChannelSender> senderFactory, Func<st
             config.GetSettings().Set("GatewayChannelSenderFactory", senderFactory);
             config.GetSettings().Set("GatewayChannelReceiverFactory", receiverFactory);
         }
+
+
+        /// <summary>
+        /// Set the number of retries and time increase between them for messages failing to be sent through the gateway.
+        /// </summary>
+        /// <param name="numberOfRetries">The total number of retries to do. 0 means no retry.</param>
```
Comment (timbussmann) [inline src/NServiceBus.Gateway/Settings/GatewaySettings.cs]: wouldn't 1 also mean no retry currently?

File: src/NServiceBus.Gateway/Settings/GatewaySettings.cs [hunk 14]
```diff
@@ -36,5 +37,38 @@ public void ChannelFactories(Func<string, IChannelSender> senderFactory, Func<st
...
+        /// <param name="timeIncrease">The time to wait between each retry.</param>
+        public void Retries(int numberOfRetries, TimeSpan timeIncrease)
+        {
+            SetDefaultRetryPolicySettings(numberOfRetries, timeIncrease);
+        }
+
+        /// <summary>
+        /// Set a retry policy that returns a TimeSpan to delay between attempts based on the number of retries attempted. Return TimeSpan.Zero to abort retries.
```
Comment (timbussmann) [inline src/NServiceBus.Gateway/Settings/GatewaySettings.cs]: can we wrap `TimeSpan.Zero` in some <see cref="TimeSpan.Zero" /> ?

File: src/NServiceBus.Gateway.AcceptanceTests/NServiceBus.Gateway.AcceptanceTests.csproj (hunk 2 above)
Comment (timbussmann) [inline src/NServiceBus.Gateway.AcceptanceTests/NServiceBus.Gateway.AcceptanceTests.csproj]: ah, ok then

**Comment by boblangley** [comment]
I am working on the maintainer suggestions.

File: src/NServiceBus.Gateway/Settings/GatewaySettings.cs (hunk 13 above)
Comment (mikeminutillo) [inline src/NServiceBus.Gateway/Settings/GatewaySettings.cs]: No . We pass current retry number into the custom policy and the `DefaultRetryPolicy` checks to see if that number is bigger this one. So if you call this with `1` (the very first retry) will be `if(1 > 1) no retry else 1 * timeIncrease`. I'll see if I can add some unit tests for the default retry policy.

Comment (boblangley) [inline src/NServiceBus.Gateway/Settings/GatewaySettings.cs]: Should it be currentRetry >= numOfRetries?

**Comment by mikeminutillo** [comment]
@Particular/nservicebus-maintainers All review comments have been addressed. Can we have a second review please?

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+            public bool CustomRetryPolicyWasCalled { get; set; }
+        }
+
+        class Headquarters : EndpointConfigurationBuilder
+        {
+            public Headquarters()
+            {
+                EndpointSetup<DefaultServer>(c =>
+                {
+                    c.ScaleOut().InstanceDiscriminator("1");
```
Comment (andreasohlund) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs]: Why is this needed?

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+        {
+            public ErrorSpy()
+            {
+                EndpointSetup<ErrorQueueSpyServer>()
+                    .CustomEndpointName(ErrorSpyQueueName);
+            }
+
+            class ErrorMessageHandler : IHandleMessages<AnyMessage>
+            {
+                public Context Context { get; set; }
```
Comment (andreasohlund) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs]: Use ctor injection instead (we're heading towards not supporting property injection ootb)=

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_without_retries.cs [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+            public int NumberOfRetries { get; set; }
+        }
+
+        class Headquarters : EndpointConfigurationBuilder
+        {
+            public Headquarters()
+            {
+                EndpointSetup<DefaultServer>(c =>
+                {
+                    c.ScaleOut().InstanceDiscriminator("1");
```
Comment (andreasohlund) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_without_retries.cs]: Is this needed?

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_without_retries.cs [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+        {
+            public ErrorSpy()
+            {
+                EndpointSetup<ErrorQueueSpyServer>()
+                    .CustomEndpointName(ErrorSpyQueueName);
+            }
+
+            class ErrorMessageHandler : IHandleMessages<AnyMessage>
+            {
+                public Context Context { get; set; }
```
Comment (andreasohlund) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_without_retries.cs]: Ctor injection

File: src/NServiceBus.Gateway.Tests/APIApprovals.Approve.approved.txt [hunk 19]
```diff
@@ -160,5 +160,8 @@ namespace NServiceBus.Settings
     public class GatewaySettings
     {
         public void ChannelFactories(System.Func<string, NServiceBus.Gateway.IChannelSender> senderFactory, System.Func<string, NServiceBus.Gateway.IChannelSender> receiverFactory) { }
+        public void CustomRetryPolicy(System.Func<NServiceBus.Transports.IncomingMessage, System.Exception, int, System.TimeSpan> customRetryPolicy) { }
```
Comment (andreasohlund) [inline src/NServiceBus.Gateway.Tests/APIApprovals.Approve.approved.txt]: :heart: that we now have a specific retry api for the GW

File: src/NServiceBus.Gateway/Gateway.cs [hunk 20]
```diff
@@ -35,7 +35,7 @@ public class Gateway : Feature
     {
         internal Gateway()
         {
-
+            DependsOn("DelayedDelivery");
```
Comment (andreasohlund) [inline src/NServiceBus.Gateway/Gateway.cs]: What happens when the GW is explicitly enabled the there `DelayedDelivery` feature is disabled? (do we throw a meaningful ex?)

File: src/NServiceBus.Gateway/Recoverability/ForwardFailedGatewayMessagesToErrorQueueBehavior.cs [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+            {
+                await next().ConfigureAwait(false);
+            }
+            catch (Exception exception)
+            {
+                await ForwardToErrorQueue(context, exception, fork).ConfigureAwait(false);
+            }
+        }
+
+        private async Task ForwardToErrorQueue(ITransportReceiveContext context, Exception exception, Func<IFaultContext, Task> fork)
```
Comment (andreasohlund) [inline src/NServiceBus.Gateway/Recoverability/ForwardFailedGatewayMessagesToErrorQueueBehavior.cs]: Redundant private

File: src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+            {
+                var sentForRetry = await SendForRetry(context, exception, fork).ConfigureAwait(false);
+                if (!sentForRetry)
+                {
+                    throw;
+                }
+            }
+        }
+
+        private async Task<bool> SendForRetry(ITransportReceiveContext context, Exception exception, Func<IRoutingContext, Task> fork)
```
Comment (andreasohlund) [inline src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs]: Redundant private

File: src/NServiceBus.Gateway/Routing/Sites/ConfigurationBasedSiteRouter.cs [hunk 23]
```diff
@@ -4,6 +4,11 @@ namespace NServiceBus.Gateway.Routing.Sites
 
     class ConfigurationBasedSiteRouter : IRouteMessagesToSites
```
Comment (andreasohlund) [inline src/NServiceBus.Gateway/Routing/Sites/ConfigurationBasedSiteRouter.cs]: Just noting that the namespace is off (should be in the root NServiceBus.Gateway namespace). We should clean that up as part of getting the beta out?

File: src/NServiceBus.Gateway/Sending/GatewaySendBehavior.cs [hunk 24]
```diff
@@ -6,42 +6,45 @@ namespace NServiceBus.Gateway.Sending
...
     class GatewaySendBehavior : PipelineTerminator<ISatelliteProcessingContext>
     {
         readonly IManageReceiveChannels channelManager;
         readonly IDispatchMessages dispatcher;
         readonly ReadOnlySettings settings;
         readonly SingleCallChannelForwarder forwarder;
-        readonly IEnumerable<IRouteMessagesToSites> routers;
+        readonly ConfigurationBasedSiteRouter configRouter;
         readonly MessageNotifier messageNotifier;
         readonly string inputAddress;
```
Comment (andreasohlund) [inline src/NServiceBus.Gateway/Sending/GatewaySendBehavior.cs]: The fields should be at the bottom of the file but I think its better to run a code cleanup as a separate pull after we get this one in so lets leave this as is for now?

File: src/NServiceBus.Gateway/Settings/GatewaySettings.cs [hunk 25]
```diff
@@ -36,5 +37,48 @@ public void ChannelFactories(Func<string, IChannelSender> senderFactory, Func<st
...
             config.GetSettings().Set("GatewayChannelReceiverFactory", receiverFactory);
         }
+
+
+        /// <summary>
+        /// Set the number of retries and time increase between them for messages failing to be sent through the gateway.
+        /// </summary>
+        /// <param name="numberOfRetries">The total number of retries to do. 0 means no retry.</param>
+        /// <param name="timeIncrease">The time to wait between each retry.</param>
+        public void Retries(int numberOfRetries, TimeSpan timeIncrease)
```
Comment (andreasohlund) [inline src/NServiceBus.Gateway/Settings/GatewaySettings.cs]: The namespace is wrong, I think all the config extensions should be in the `NServiceBus` namespace? 

Or do we use `NServiceBus.Gateway` ?

Comment (andreasohlund) [inline src/NServiceBus.Gateway/Settings/GatewaySettings.cs]: Talking to @SimonCropp and @timbussmann we propose that we go with `NServiceBus`

File: src/NServiceBus.Gateway/Settings/GatewaySettings.cs (hunk 13 above)
Comment (timbussmann) [inline src/NServiceBus.Gateway/Settings/GatewaySettings.cs]: >  So if you call this with 1 (the very first retry) will be if(1 > 1) no retry else 1 * timeIncrease. I'll see if I can add some unit tests for the default retry policy

at the time of my comment, the retry number on the first failed processing (not a retry) was already 1, that's why I asked. Seems to be fixed now.

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs [hunk 26]
```diff
@@ -0,0 +1,161 @@
...
+    {
+        [Test]
+        public async Task Should_have_been_retried()
+        {
+            var context = await Scenario.Define<Context>(c => { c.Id = Guid.NewGuid(); })
+                .WithEndpoint<Headquarters>(b =>
+                {
+                    b.CustomConfig(c => c.Gateway().Retries(3, TimeSpan.FromSeconds(1)));
+
+                    b.When((bus, c) => bus.SendToSites(new[]
```
Comment (timbussmann) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs]: when can be chained after CustomConfig, no need for a statement lambda

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs [hunk 27]
```diff
@@ -0,0 +1,161 @@
...
+                        return TimeSpan.MinValue;
+                    }));
+                    b.When((bus, ctx) => bus.SendToSites(new[]
+                    {
+                        "SiteA"
+                    }, new AnyMessage
+                    {
+                        Id = ctx.Id
+                    }));
+                })
```
Comment (timbussmann) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs]: same comment as above

File: src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs [hunk 28]
```diff
@@ -0,0 +1,161 @@
...
+
+                public Task Handle(AnyMessage errorMessage, IMessageHandlerContext context)
+                {
+                    if (errorMessage.Id == Context.Id)
+                    {
+                        if (context.MessageHeaders.ContainsKey(Headers.Retries))
+                        {
+                            Context.NumberOfRetries = Parse(context.MessageHeaders[Headers.Retries]);
+                        }
+                        Context.MessageMovedToErrorQueue = true;
```
Comment (timbussmann) [inline src/NServiceBus.Gateway.AcceptanceTests/When_sending_fails_with_retries.cs]: is this needed in the Headquaters message handler?

File: src/NServiceBus.Gateway/GatewayHttpListenerInstaller.cs [hunk 29]
```diff
@@ -92,11 +92,11 @@ static void StartNetshProcess(string identity, Uri uri)
                 }
                 var error = process.StandardOutput.ReadToEnd().Trim();
                 var message = string.Format(
-@"Failed to grant to grant user '{0}' HttpListener permissions. Processing will continue. 
+@"Failed to grant to grant user '{0}' HttpListener permissions. Processing will continue.
```
Comment (timbussmann) [inline src/NServiceBus.Gateway/GatewayHttpListenerInstaller.cs]: to grant to grant?

File: src/NServiceBus.Gateway/GatewayHttpListenerInstaller.cs [hunk 30]
```diff
@@ -60,7 +60,7 @@ public Task Install(string identity)
                 catch (Exception exception)
                 {
                     var message = string.Format(
-@"Failed to grant to grant user '{0}' HttpListener permissions due to an Exception. Processing will continue.  
+@"Failed to grant to grant user '{0}' HttpListener permissions due to an Exception. Processing will continue.
```
Comment (timbussmann) [inline src/NServiceBus.Gateway/GatewayHttpListenerInstaller.cs]: to grant to grant?

File: src/NServiceBus.Gateway/Recoverability/DefaultRetryPolicy.cs [hunk 31]
```diff
@@ -0,0 +1,20 @@
...
+    class DefaultRetryPolicy
+    {
+        public static Func<IncomingMessage, Exception, int, TimeSpan> BuildWithDefaults()
+        {
+            return Build(4, TimeSpan.FromSeconds(60));
+        }
+
+        public static Func<IncomingMessage, Exception, int, TimeSpan> Build(int numberOfRetries, TimeSpan timeIncrease)
+        {
+            return (message, exception, currentRetry) => currentRetry > numberOfRetries
```
Comment (timbussmann) [inline src/NServiceBus.Gateway/Recoverability/DefaultRetryPolicy.cs]: it seems to me like thi should be `currentRetry >= numberOfRetries`.
e.g. the default policy configures for 4 retries, but it will actually do 3 retries as stated by the acceptance test.

File: src/NServiceBus.Gateway/Recoverability/ForwardFailedGatewayMessagesToErrorQueueBehavior.cs [hunk 32]
```diff
@@ -0,0 +1,54 @@
...
+        {
+            try
+            {
+                var message = context.Message;
+                Logger.Error($"Moving gateway message '{message.MessageId}' from '{localAddress}' to the error queue because processing failed due to an exception:", exception);
```
Comment (timbussmann) [inline src/NServiceBus.Gateway/Recoverability/ForwardFailedGatewayMessagesToErrorQueueBehavior.cs]: this should use `ErrorFormat`

File: src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs [hunk 33]
```diff
@@ -0,0 +1,99 @@
...
+            Logger.Warn($"Gateway message '{message.MessageId}' failed. Will reschedule message after {timeIncrease}:", exception);
+
+            var messageToRetry = new OutgoingMessage(message.MessageId, message.Headers, message.Body);
+
+            messageToRetry.Headers[Headers.Retries] = currentRetry.ToString();
+            messageToRetry.Headers[Headers.RetriesTimestamp] = DateTimeExtensions.ToWireFormattedString(DateTime.UtcNow);
+
+            var dispatchContext = this.CreateRoutingContext(messageToRetry, localAddress, context);
+
+            context.Extensions.Set(new List<DeliveryConstraint>
```
Comment (timbussmann) [inline src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs]: shouldn't that be `dispatchContext`?

File: src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs [hunk 34]
```diff
@@ -0,0 +1,99 @@
...
+            var currentRetry = retryCount + 1;
+
+            var timeIncrease = retryPolicy(message, exception, currentRetry);
+
+            if (timeIncrease <= TimeSpan.MinValue)
+            {
+                return false;
+            }
+
+            Logger.Warn($"Gateway message '{message.MessageId}' failed. Will reschedule message after {timeIncrease}:", exception);
```
Comment (timbussmann) [inline src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs]: use WarnFormat

File: src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs [hunk 35]
```diff
@@ -0,0 +1,99 @@
...
+                if (int.TryParse(value, out i))
+                {
+                    return i;
+                }
+            }
+            return 0;
+        }
+
+        string localAddress;
+        readonly Func<IncomingMessage, Exception, int, TimeSpan> retryPolicy;
```
Comment (timbussmann) [inline src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs]: remove readonly

File: src/NServiceBus.Gateway/Gateway.cs [hunk 36]
```diff
@@ -62,7 +62,12 @@ protected override void Setup(FeatureConfigurationContext context)
             Func<string, IChannelReceiver> channelReceiverFactory;
             RegisterChannels(context, channelManager, out channelSenderFactory, out channelReceiverFactory);
 
-            gatewayPipeline.Register("GatewaySendProcessor", b => new GatewaySendBehavior(gatewayInputAddress, channelManager, new MessageNotifier(), b.Build<IDispatchMessages>(), context.Settings, CreateForwarder(channelSenderFactory, b.BuildAll<IDataBus>()?.FirstOrDefault()), CreateSiteRouters(context)), "Processes messages to be sent to the gateway");
+            gatewayPipeline.Register("ForwardFailedGatewayMessagesToErrorQueue", b => new ForwardFailedGatewayMessagesToErrorQueueBehavior(gatewayInputAddress, b.Build<CriticalError>()), "Handles failures when sending messages through the gateway");
+
+            var retryPolicy = context.Settings.HasSetting("Gateway.Retries.RetryPolicy") ? context.Settings.Get<Func<IncomingMessage, Exception, int ,TimeSpan>>("Gateway.Retries.RetryPolicy") : DefaultRetryPolicy.BuildWithDefaults();
+
+            gatewayPipeline.Register(new RetryFailedGatewayMessagesBehavior.Registration(gatewayInputAddress, retryPolicy, "ForwardFailedGatewayMessagesToErrorQueue"));
```
Comment (timbussmann) [inline src/NServiceBus.Gateway/Gateway.cs]: passing in the dependency as a parameter but duplicating the string instead of using a variable seems weird.
Why not provide a constant for `"ForwardFailedGatewayMessagesToErrorQueue"` and reference that directly without passing it as a parameter?

File: src/NServiceBus.Gateway/Routing/Sites/ConfigurationBasedSiteRouter.cs [hunk 37]
```diff
@@ -14,14 +19,14 @@ public IEnumerable<Site> GetDestinationSitesFor(Dictionary<string, string> heade
...
+                    if (sites.TryGetValue(siteKey, out site))
                     {
                         yield return site;
                     }
                 }
             }
         }
 
-        public IDictionary<string, Site> Sites { get; set; }
+        readonly IDictionary<string, Site> sites;
```
Comment (timbussmann) [inline src/NServiceBus.Gateway/Routing/Sites/ConfigurationBasedSiteRouter.cs]: remove readonly

File: src/NServiceBus.Gateway/Settings/GatewaySettings.cs (hunk 25 above)
Comment (boblangley) [inline src/NServiceBus.Gateway/Settings/GatewaySettings.cs]: Namespace changed

File: src/NServiceBus.Gateway/Recoverability/DefaultRetryPolicy.cs (hunk 31 above)
Comment (boblangley) [inline src/NServiceBus.Gateway/Recoverability/DefaultRetryPolicy.cs]: Working as intended, see above / acceptance tests

File: src/NServiceBus.Gateway/Recoverability/ForwardFailedGatewayMessagesToErrorQueueBehavior.cs (hunk 32 above)
Comment (boblangley) [inline src/NServiceBus.Gateway/Recoverability/ForwardFailedGatewayMessagesToErrorQueueBehavior.cs]: ErrorFormat doesn't accept an Exception.

File: src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs (hunk 34 above)
Comment (boblangley) [inline src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs]: WarnFormat doesn't accept an Exception.

File: src/NServiceBus.Gateway/Routing/Sites/ConfigurationBasedSiteRouter.cs (hunk 23 above)
Comment (boblangley) [inline src/NServiceBus.Gateway/Routing/Sites/ConfigurationBasedSiteRouter.cs]: There are other files that are using similar namespace paths. Should we fix all of them?

Comment (andreasohlund) [inline src/NServiceBus.Gateway/Routing/Sites/ConfigurationBasedSiteRouter.cs]: Let's deal with those separately as part of preparing the GW for beta/rc?

File: src/NServiceBus.Gateway/Recoverability/ForwardFailedGatewayMessagesToErrorQueueBehavior.cs (hunk 32 above)
Comment (timbussmann) [inline src/NServiceBus.Gateway/Recoverability/ForwardFailedGatewayMessagesToErrorQueueBehavior.cs]: oh alright then. Seems like we should add that?

**Comment by timbussmann** [comment]
LGTM. Title is still [WIP]. Since @andreasohlund mentioned beta-2 packages being available, do you want to bake update them in this PR or do you want to raise a new PR?

**Comment by boblangley** [comment]
Ready to merge, waiting for build server to agree.
